### PR TITLE
iio: add iio core

### PR DESCRIPTION
**Pull request by cristiantpop**
This module implements read/write ops, required by libtinyiiod and further
calls show/store functions, corresponding to device/channel/attribute.

Signed-off-by: Cristian Pop <cristian.pop@analog.com>

### CODE REVIEW

File: iio/iio.c [hunk 1]
```diff
@@ -0,0 +1,765 @@
...
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+struct iio_interface {
+	const char *name;
+	uint16_t num_channels;
+	uint32_t ch_mask;
+	void *dev_instance;
+	struct iio_device *iio;
+	ssize_t (*get_device_xml)(char** xml, const char *device_name, uint8_t ch_no);
```
Comment (mirceacaprioru) [inline iio/iio.c]: should it be char **xml ? for consistency ?

File: iio/iio.c [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+	if (!(*channels))
+		return -EINVAL;
+
+	while (channels[i]) {
+		if (!strcmp(channel, channels[i]->name))
+			return i;
+		i++;
+	}
+
+	return -ENODEV;
```
Comment (mirceacaprioru) [inline iio/iio.c]: i think ENOENT fits better here.

File: iio/iio.c [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+	if (!(*attributes))
+		return -EINVAL;
+
+	while (attributes[i]) {
+		if (!strcmp(attr, attributes[i]->name))
+			return i;
+		i++;
+	}
+
+	return -ENODEV;
```
Comment (mirceacaprioru) [inline iio/iio.c]: same as above.

File: iio/iio.c [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+
+static struct iio_interface *iio_get_interface(const char *device_name,
+		struct iio_interfaces *iio_interfaces)
+{
+	int16_t i = 0;
+
+	for (i = 0; i < iio_interfaces->num_interfaces; i++) {
+		if (!strcmp(device_name, iio_interfaces->interfaces[i]->name)) {
+			return iio_interfaces->interfaces[i];
+		}
```
Comment (mirceacaprioru) [inline iio/iio.c]: these brackets can be removed.

File: iio/iio.c [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+ * @param *channel channel properties
+ * @param *map map with coresponding attributes
+ * @param map_size
+ * @return length of chars written in buf
+ */
+static ssize_t iio_read_all_attr(void *device, char *buf, size_t len,
+				 const struct iio_ch_info *channel, struct iio_attribute **attributes)
+{
+	int16_t i = 0, j = 0;
+	char local_buf[256];
```
Comment (mirceacaprioru) [inline iio/iio.c]: should this be initialized to 0 ?

Comment (cristiantpop) [reply iio/iio.c]: Not necessary.

File: iio/iio.c (hunk 4 above)
Comment (cristiantpop) [reply iio/iio.c]: OK.

File: iio/iio.c (hunk 3 above)
Comment (cristiantpop) [reply iio/iio.c]: OK

File: iio/iio.c (hunk 2 above)
Comment (cristiantpop) [reply iio/iio.c]: OK

File: iio/iio.c (hunk 1 above)
Comment (cristiantpop) [reply iio/iio.c]: OK

**Comment by cristiantpop** [comment]
V2:

- Remove unnecessary brackets.

- Return "ENOENT".

- declare "char **xml", for consistency.

**Comment by cristiantpop** [comment]
V3:

- Add some missing function headers.

**Comment by cristiantpop** [comment]
V4:

- call "iio_axi_adc_transfer_dev_to_mem" with "iiod_adc" parameter that contains "dcache_invalidate_range".

File: iio/iio.c [hunk 6]
```diff
@@ -0,0 +1,787 @@
...
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include "iio.h"
+#include "ctype.h"
+#include "tinyiiod.h"
+#include "util.h"
+#include "error.h"
+#include "iio_axi_adc.h"
+#include "iio_axi_dac.h"
```
Comment (dbogdan) [inline iio/iio.c]: I thought iio.c is platform agnostic and generic... including iio_axi_adc.h and iio_axi_dac.h turns out it is not.

**Comment by cristiantpop** [comment]
V5:

- Remove dependency of "iio_axi_dac.h" and "iio_axi_adc.h".

- Refactor prototype for "iio_register" to use "struct iio_interface_init_par".

- Refactor "iio_transfer_dev_to_mem", "iio_read_dev", "iio_transfer_mem_to_dev", "iio_write_dev" to use callbacks if they exist.

File: iio/iio.c [hunk 7]
```diff
@@ -0,0 +1,797 @@
...
+///**
+// * write channel attribute
+// * @param *device name
+// * @param *channel name
+// * @param *ch_out type: input/output
+// * @param *attr name
+// * @param *buff where value is stored
+// * @param len length of the value
+// * @return length of chars written to attribute, negative value in case of failure
+// */
```
Comment (dbogdan) [inline iio/iio.c]: For consistency with the rest of the file, can you update these comments?

File: iio/iio.c [hunk 8]
```diff
@@ -0,0 +1,797 @@
...
+	return NULL;
+}
+
+/**
+ * read all attributes from map
+ * @param *buff where value is stored
+ * @param len maximum length of value to be stored in buf
+ * @param *channel channel properties
+ * @param *map map with coresponding attributes
+ * @param map_size
```
Comment (dbogdan) [inline iio/iio.c]: This parameter is not present.

File: iio/iio.c [hunk 9]
```diff
@@ -0,0 +1,797 @@
...
+	if(iio_interface->write_data)
+		return iio_interface->write_data(iio_interface->dev_instance, (char*)buf,
+						 offset, bytes_count);
+
+	return -ENOENT;
+}
+
+/**
+ * @brief Get a merged xml containing all devices.
+ * @param devs - physical device instance.
```
Comment (dbogdan) [inline iio/iio.c]: Missing parameter - can you check again all the comments and add proper description for all parameters?

File: iio/iio.c [hunk 10]
```diff
@@ -0,0 +1,797 @@
...
+
+	return -ENOENT;
+}
+
+/**
+ * Check device
+ * @param *device
+ * @return TRUE if valid device is found
+ */
+static bool iio_supporter_dev(const char *device)
```
Comment (mirceacaprioru) [inline iio/iio.c]: i think iio_supported_dev would fit better

**Comment by cristiantpop** [comment]
V6:

- Rename to: "iio_supported_dev".

- fix parameters comments.

**Pull request closed without merging**